examples/algorithms/cuda/include/cuda/track_finding/track_params_estimation_algorithm.hpp

```diff
@@ -0,0 +1,39 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include "cuda/seeding/track_params_estimation.hpp"
+
+namespace traccc {
+namespace cuda {
+
+class track_params_estimation_algorithm {
+
+    public:
+    using input_type =
+        std::pair<host_measurement_container&, host_seed_container&>;
+    using output_type = host_bound_track_parameters_collection;
+
+    track_params_estimation_algorithm(vecmem::memory_resource* mr)
+        : params(mr) {
+        tp = std::make_shared<traccc::cuda::track_params_estimation>(
+            traccc::cuda::track_params_estimation(mr));
+    }
+
+    output_type& operator()(input_type& i) {
+        params = tp->operator()(i);
+        return params;
+    }
+
+    private:
+    output_type params;
+    std::shared_ptr<traccc::cuda::track_params_estimation> tp;
```

Comment:
Is it really necessary to use a shared pointer here? Remember, these things are not cheap.

Reply:
I don't think this is necessarily a problem here; this is a pointer to a function object, so it is unlikely to be replaced very often. Since `std::shared_ptr` has zero de-referencing overhead, I don't foresee any performance degradation here.

Whether it is really necessary is debatable, but I can see some benefit to this approach given that function objects in C++ are often assumed to be copyable. Using shared pointers here can feasibly reduce the overhead of that copying. But perhaps Beomki had different reasons in mind?

Reply:
@stephenswat is right - This algorithm objects are generated only once before the actual algorithm runs. It could also be normal objects but I was too lazy to make a copy operator for algorithm classes. And if it doesn't introduce any overheads, we can just keep this way.

Reply:
Just as for `traccc::track_params_estimation_algorithm`, I don't see what's the purpose of this class. :confused: What interface does it add on top of `traccc::cuda::track_params_estimation`?

Also, why does it not inherit from the `traccc::algorithm` base class? :confused: